### PR TITLE
FIX-#2442: fixed Series assignment with different indices

### LINKED ISSUE #2442: TypeError in case of attempt of appending new column when frame has duplicated timestamp indices
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any
- **Modin version** (`modin.__version__`): 3e32d0273e56c192b104fb65ad8fcbcdc801647c
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
import modin.pandas as pd
import pandas
import numpy as np

data = {"a": np.arange(4)}

# Index with duplicated timestamp
index = pd.to_datetime(["2018-06-05", "2018-06-05", "2018-06-07", "2018-06-08"])

md_df, pd_df = pd.DataFrame(data, index=index), pandas.DataFrame(data, index=index)

pd_df["b"] = pandas.Series(np.zeros(len(pd_df))) # Works fine
md_df["c"] = np.zeros(len(md_df)) # Works fine
md_df["b"] = pd.Series(np.zeros(len(md_df))) # TypeError
```

<details><summary> Traceback </summary>

```
Traceback (most recent call last):
  File "test_outer.py", line 13, in <module>
    md_df["b"] = pd.Series(np.zeros(len(md_df))) # TypeError
  File "/localdisk/dchigare/repos/modin_bp/modin/pandas/dataframe.py", line 1978, in __setitem__
    join="left",
  File "/localdisk/dchigare/repos/modin_bp/modin/backends/pandas/query_compiler.py", line 303, in concat
    new_modin_frame = self._modin_frame._concat(axis, other_modin_frame, join, sort)
  File "/localdisk/dchigare/repos/modin_bp/modin/engines/base/frame/data.py", line 1850, in _concat
    axis ^ 1, others, how, sort, force_repartition=True
  File "/localdisk/dchigare/repos/modin_bp/modin/engines/base/frame/data.py", line 1701, in _copartition
    joined_index = self._join_index_objects(axis, index_other_obj, how, sort)
  File "/localdisk/dchigare/repos/modin_bp/modin/engines/base/frame/data.py", line 980, in _join_index_objects
    joined_obj = merge_index(joined_obj, obj)
  File "/localdisk/dchigare/repos/modin_bp/modin/engines/base/frame/data.py", line 974, in merge_index
    return obj1.join(obj2, how=how, sort=sort)
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/indexes/datetimelike.py", line 893, in join
    sort=sort,
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/indexes/base.py", line 3483, in join
    return this.join(other, how=how, return_indexers=return_indexers)
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/indexes/base.py", line 3494, in join
    other, how=how, return_indexers=return_indexers
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/indexes/base.py", line 3815, in _join_monotonic
    join_index, lidx, ridx = self._left_indexer(sv, ov)
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/indexes/base.py", line 257, in _left_indexer
    return libjoin.left_join_indexer(left, right)
  File "pandas/_libs/join.pyx", line 357, in pandas._libs.join.left_join_indexer
TypeError: '<' not supported between instances of 'Timestamp' and 'int'

```

</details>

### Describe the problem

The problem appears only if we're trying to assign modin Series, because if value to assign has query compiler then we will go in the problem branch where we will try to concat `self` frame with passed Series:
https://github.com/modin-project/modin/blob/3e32d0273e56c192b104fb65ad8fcbcdc801647c/modin/pandas/dataframe.py#L1970-L1981

The problem is that inside `qc::concat->modin_frame::_concat->modin_frame::_copartition` index joining happens
https://github.com/modin-project/modin/blob/3e32d0273e56c192b104fb65ad8fcbcdc801647c/modin/engines/base/frame/data.py#L1750

Which cannot be performed between non-unique TimeStamp and integer index, so we're failing here.

### How pandas handle that case?
In pandas `insert` is separate from `concat/join/union` function, so it does not joining indices at all. It goes from `insert->_sanitize_column` that handles index mismatching by [reindexing value to self index](https://github.com/pandas-dev/pandas/blob/c77fc357efbe9443ad8e32b9ce34599e61bcd3f5/pandas/core/frame.py#L3793-L3813).

### Proposal to fix
Replace calling `query_compiler._concat` by `DataFrame.insert` which will be responsible to match indices without joining it

### Temporal workarounds
Since the problem appears only with those values that has query compiler, all we need to do is to pass a non-modin object as a value to assign

```python
try:
    df["new_col1"] = pd.Series(np.arange(5))
except ValueError:
    # workaround for #2442
    df["new_col1"] = np.arange(5)

try:
    df["new_col2"] = some_series
except ValueError:
    # workaround for #2442
    df["new_col2"] = some_series.values
```

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/CONTRIBUTING.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2442 #1772
- [x] tests added and passing

## What was done
**Pandas backend:**
- Added `modin/test/backends/base` folder to test methods of all query compilers
- Moved content of `modin/pandas/test/conftest.py` to `modin/conftest.py` to `--backend` parameter will be available in internal tests
- Added ability at `insert` to process modin objects without translating it into pandas
- Added `insert_item` method to a query compiler level, that encapsulates logic of `setitem` and `insert`
- For pandas engines `how=None` parameter was added to `modin_frame._concat`, which indicates that `joined_index` should be considered as the index of `self`. Passing `how=None` in case of `insert` and `setitem`.
- Reworked `_copartition` to be able to repartition when `self` frame is empty: now base frame (for what partition lengths other partitions should align) considered to be the first that is not empty

**OmniSci backend:**
- It also got an ability to `insert` modin Series, but only those, which parent is `self` frame (limitations of OmniSci backend `_concat`), otherwise it defaults to pandas

### CODE REVIEW

**Comment by anmyachev** [comment]
@dchigarev 

What about next test? Does the PR also fix it?

```python
def test_aligning_partitions():
    data = [0, 1, 2, 3, 4, 5]
    modin_df1, _ = create_test_dfs({"a": data, "b": data})
    modin_df = modin_df1.loc[:2]

    modin_df2 = modin_df.append(modin_df)

    modin_df2["c"] = modin_df1["b"]
    repr(modin_df2)
```

**Comment by dchigarev** [comment]
Test above fails in the current branch, however after merging #2378 into #2443 that test passes. So it seems that for now that PR is blocked by #2378

**Comment by YarShev** [comment]
#2378 was merged.

**Comment by dchigarev** [comment]
ASV benchmarks:
```
asv continuous src/master HEAD -b TimeInsert -b TimeSetItem -b TimeConcat -b TimeBinaryOp --launch-method=spawn

 before           after         ratio
     [031f444f]       [a177996f]
     <src/master>       <issue-2442>
-         144±3ms          119±2ms     0.83  benchmarks.TimeInsert.time_insert_qc('int', (10000, 128), 1, 'zero', False)
-         116±3ms         93.4±2ms     0.81  benchmarks.TimeInsert.time_insert_qc('int', (10000, 128), 1, 'zero', True)
-         160±2ms          124±2ms     0.78  benchmarks.TimeSetItem.time_setitem_qc('int', (10000, 128), 1, 'zero', False)
-         113±2ms       76.7±0.5ms     0.68  benchmarks.TimeInsert.time_insert_qc('int', (10000, 128), 1, 'middle', True)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

File: .github/workflows/ci.yml
Comment (anmyachev) [inline .github/workflows/ci.yml:232]: In other cases, we control the backend through an environment variable.

Why is `--backend` pytest option now used?

Comment (dchigarev) [reply .github/workflows/ci.yml:232]: This is the way how we're running tests for the base backend, there is no backend called "base", because `BaseQueryCompiler` is an abstract class. We dynamically generate "base" backend just to run tests, by replacing abstract methods of `BaseQueryCompiler` via `PandasQueryCompilers`s methods. This logic was introduced in #2047

File: modin/backends/pandas/query_compiler.py
Comment (anmyachev) [inline modin/backends/pandas/query_compiler.py:2171]: Missing description about `how` and `replace` parameters.

File: modin/backends/pandas/query_compiler.py
Comment (anmyachev) [inline modin/backends/pandas/query_compiler.py:2198]: Why we need `+ int(replace)` here?

File: modin/engines/base/frame/data.py
Comment (anmyachev) [inline modin/engines/base/frame/data.py:1034]: Why moved?

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (anmyachev) [inline modin/experimental/backends/omnisci/query_compiler.py:522]: There is no explicit check for whether `value` is a subframe or not?

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:522]: Engine check it inside and raises `NotImplementedError` if it is not a subframe, in this case we're catching this exception and defaulting to pandas

File: asv_bench/benchmarks/benchmarks.py
Comment (dchigarev) [reply asv_bench/benchmarks/benchmarks.py:210]: to trigger materialization, or it is not included in benchmarks?

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:522]: Technically we can do this explicit check at the query compiler level by calling `modin_frame._find_common_projection_base` (this is how the engine determines if the other frame is a subframe of `self`), but this method is linear on the length of operation tree and gives us some overhead of inserting because we will call this method twice: first in query compiler and then in the engine.

discussed it with @ienkovich and agreed on an approach with catching exception

Comment (anmyachev) [reply modin/experimental/backends/omnisci/query_compiler.py:522]: ok

File: asv_bench/benchmarks/benchmarks.py
Comment (anmyachev) [reply asv_bench/benchmarks/benchmarks.py:210]: Maybe another way will bring less overhead to measuring time of setitem. What about using `shape`?

File: modin/test/backends/base/test_internals.py
Comment (anmyachev) [inline modin/test/backends/base/test_internals.py:67]: What type of `md_frames` in case of using `post_fn` parameter?

File: modin/pandas/test/dataframe/test_map_metadata.py
Comment (anmyachev) [inline modin/pandas/test/dataframe/test_map_metadata.py:931]: This is fixed?

File: modin/pandas/dataframe.py
Comment (anmyachev) [inline modin/pandas/dataframe.py:1994]: Why removed?

File: asv_bench/benchmarks/benchmarks.py
Comment (dchigarev) [reply asv_bench/benchmarks/benchmarks.py]: good idea, applied it

File: modin/backends/pandas/query_compiler.py
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py]: added

File: modin/backends/pandas/query_compiler.py
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py]: for nothing, removed it, thanks

File: modin/engines/base/frame/data.py
Comment (dchigarev) [reply modin/engines/base/frame/data.py:1034]: if `do_join_index is False` but `is_duplicates is True` we want to get reindexers, so moved it under `if do_join_index` branch to be available in cases when we don't want to join indices

File: modin/pandas/test/dataframe/test_map_metadata.py
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_map_metadata.py:931]: it seems, I've removed `xfail` decorator and it passed)

File: modin/test/backends/base/test_internals.py
Comment (dchigarev) [reply modin/test/backends/base/test_internals.py:67]: it depends on what `post_fn` returns, in our case it will be a tuple of two modin frames

File: modin/pandas/dataframe.py
Comment (dchigarev) [reply modin/pandas/dataframe.py:1994]: now in this case is processed by `DataFrame.insert`

File: modin/engines/base/frame/data.py
Comment (anmyachev) [reply modin/engines/base/frame/data.py:1034]: Why do we need renindexers if we don't do join?

Comment (dchigarev) [reply modin/engines/base/frame/data.py:1034]: it happens when `how` parameter is `None`, in that case we're not joining indexes by considering `joined_index` as `self.index`

Comment (anmyachev) [reply modin/engines/base/frame/data.py:1034]: So more clear condition is `if is_duplicates and how is None`, isn't it?

Comment (anmyachev) [reply modin/engines/base/frame/data.py:1034]: Also note that in case when:
```python
any(not indexes[0].equals(index) for index in [indexes[1:]]) == True
```
we don't need to get indexers

Comment (dchigarev) [reply modin/engines/base/frame/data.py:1034]: >So more clear condition is if is_duplicates and how is None, isn't it?

this won't work in cases when `how is not None`. 
Why do we have a statement like that:
We want to get `indexers` when `is_duplicated and (do_join_index or how is not None)`, when `do_join_index is True` we're falling into [this branch](https://github.com/dchigarev/modin/blob/1c6b2588318960381d467d704ba0392abe05b1f4/modin/engines/base/frame/data.py#L1016) and if `len(indexes) == 2 and is_duplicates` we're computing `indexers` in a fast path, so don't want to compute them again, otherwise we should compute them further. After determining what `joined_index` is, we're hitting [this if](https://github.com/dchigarev/modin/blob/1c6b2588318960381d467d704ba0392abe05b1f4/modin/engines/base/frame/data.py#L1032) that says, that if we need indexers (`is_duplicates is True`) but didn't computed it earlier (`indexers[0] is None`) them compute indexers

Comment (dchigarev) [reply modin/engines/base/frame/data.py]: reworked it in the last commit

**Review by anmyachev — APPROVED** [review]
LGTM!

**Pull request merged**
